Add notes for Dec 2021 plenary (#175)
tc39/tc39-notes: meetings/2021-12/dec-14.md

KG: Yes, that's an excellent point.

## ECMA-402 Status

Presenter: Ujjwal Sharma (USA)

USA: In the interest of time, I'd really quickly make the statement that there is no update and so there's no new normative pull request that we need consensus on, which means that we're free for ecma 402. Thank you.

## ECMA-404 Status

RPR: Anything on Ecma 404, Chip?

CM: It's exactly what you expect. All is well, all is stable. Nothing changes. That's the way it's supposed to be.

## Test262 Status

Presenter: Rick Waldron (RW)

RW: Regular additions and updates. Frequency of contribution. It's all exactly as I would expect it to be for a large-scale repository full of conformance tests. Stage 3 stuff is doing great. If anybody has any questions, hit me up.

## Updates from the CoC committee

JHD: I don't think there's really any updates right now.

## Add import.meta[Symbol.toStringTag]

Presenter: Jordan Harband (JHD)

- [PR](https://github.com/tc39/ecma262/pull/2106)

> JHD: So, the committee has previously decided essentially that name space objects like reflection and atomics should have a @@toStringTag for debuggability.

Reviewer: ```suggestion
JHD: So, the committee has previously decided essentially that namespace objects like Reflect, JSON, and Atomics should have a `@@toStringTag` for debuggability.
```

Only really, what is the proposed content of the two string tag? Oh, in this pull request. It is the string import meta in Pascal case.

MM:  In what case Pascal-like capital?

JHD: Yes, capital I.

MM: Got it. it. Thank you.

> JHD: Yeah, and it could be anything the exact content of the string is I think pretty byte shuttle. The thing that we need consensus on most. Most importantly here is, should we add the property or not? The arguments for would be debug ability and you know, it's a built-in object that holds stuff like reflective Amendment and you know math and atomics  and so on. The arguments against it would be that, it's not like a globally available one where the same instance is always everywhere, each module gets their own so, Bradley commented that PR year and a half ago for some reasons why we may be shouldn't do it. So, the analog here, there's three bullet points. The fourth bullet point here to me, is sort of contradicted by the fact that module namespace objects already have own symbol that to string tag properties. So even though every module has can be imported as its own module name, space object, that will have a symbol to string had property. So for me, the that can import that matter are conceptually similar. There's no like, you know, one per module and The hold on Jordan get have bugged.

Reviewer: ```suggestion
JHD: Yeah, and it could be anything the exact content of the string is I think pretty bikesheddable. The thing that we need consensus on most importantly here is, should we add the property or not? The arguments "for" would be debuggability. It's a built-in object that holds stuff like Reflect, Math, Atomics, and so on. The arguments "against" would be that, it's not like a globally available one where the same instance is always everywhere, each module gets their own. Bradley commented on that PR a year and a half ago with some reasons why we maybe shouldn't do it - there's three bullet points. The fourth bullet point here to me, is sort of contradicted by the fact that module namespace objects already have own `Symbol.toStringTag` properties. So even though every module can be imported as its own module namespace object, that will have a `Symbol.toStringTag` property. So for me, that and `import.meta` are conceptually similar. There's not one per module.
```

MB: I think that's it. Happy to move on. Thanks everyone.

### Conclusion/Resolution

- proposal is ready for stage 3 reviewers
- stated non-blocking concerns about the necessity of a new flag from Mozilla

## Array Grouping for Stage 3

Presenter: Justin Ridgewell (JRL)

- [proposal](https://github.com/tc39/proposal-array-grouping)
- [criteria](https://github.com/tc39/proposal-array-grouping/issues/19)

JRL: This is super easy because there are essentially no changes. We reached the stage 3 criteria since the last meeting for Array `groupBy` and `groupByMap`. We have spec text. That's been pushed out, reviewers have all approved and editors have all approved - except one change that's currently in flight, but it's an editorial change to the methods.

> JRL: The only discussion point which needs to be brought up is, Jordan brought up a groupByMap naming issue. Essentially. He's equating flatMap meaning `map` followed by `flat` and if that would be confusing for people who see groupByMap as map followed by groupBy, or groupBy into a  map output. I'm not super sold on it. I think it's okay to have groupByMap return a map. But that's the only discussion point we have - and then I can ask for stage 3.

Reviewer: ```suggestion
JRL: The only discussion point which needs to be brought up is, JHD brought up a groupByMap naming issue. Essentially. He's equating flatMap meaning `map` followed by `flat` and if that would be confusing for people who see groupByMap as map followed by groupBy, or groupBy into a  map output. I'm not super sold on it. I think it's okay to have groupByMap return a map. But that's the only discussion point we have - and then I can ask for stage 3.
```

> JHD: Yeah, I just want to make sure we discussed this. So yeah, I mean it's a relatively minor issue. The polyfill that I already made would have to be renamed so it's not like - this will cause friction for me to rename it anyway, but I still wanted to bring it up. The other thing is - it's kind of unfortunate, right? We have this naming conflict between mapping and a Map and it seems - I don't really have a better suggestion for the name, but it seems like not a great name for the groups, where the groups are maps instead of objects. Before stage 3 is the time to discuss it. So I wanted to bring it up before the advancement.

Reviewer: ```suggestion
JHD: Yeah, I just want to make sure we discussed this. So yeah, I mean it's a relatively minor issue. The polyfill that I already made would have to be renamed so this will cause friction for me to rename it anyway, but I still wanted to bring it up. The other thing is - it's kind of unfortunate, right? We have this naming conflict between mapping and a Map and it seems - I don't really have a better suggestion for the name, but it seems like not a great name for the groups, where the groups are Maps instead of Objects. Before stage 3 is the time to discuss it. So I wanted to bring it up before the advancement.
```

MM: right problem with the constructors is you can't have multiple Constructors where the constructors differ a Constructor naming pattern,

KG: Anyway fromEntries certainly does not take an array-like I would think that only specifically, these two methods, the array from and array.fromAsync should take array-likes. The general precedent in the language is to work with the iterable protocol.

MM: Also Constructors cannot return a promise. So with all of these things raised, I'm completely on board with this proposal should only deal with the ray from but and I'm Glad we're thinking about the others. I think that it's something we could actually consider as part of that overall future consideration. Whether there should a set.from and set.fromAsync and a map, got from entries and mapped from entries eyesight. Not saying I'm in favor of that, but I think

JHD: There is a proposal for Set.from: <https://github.com/tc39/proposal-setmap-offrom>

JSC: Yes, there is a proposal for that.

JSC: I'm asking for Stage 2 consensus. Any blockers?

BT: We have a few explicit approvals and thumbs up.

JSC: I would like to request for a formal reviewer for Stage 3.

> JHD: I'd happy to be a reviewer.

Reviewer: ```suggestion
JHD: I'd be happy to be a reviewer.
```
